### PR TITLE
Proposed final MVP QUIC API design document

### PR DESCRIPTION
**Pull request by hlandau**
Feedback welcomed. @mattcaswell

### CODE REVIEW

File: doc/designs/quic-design/synthesis.md [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+ * Get the polling FD for the BIO. This might be different from the value
+ * returned by BIO_get_fd and should be used by the application for polling. For
+ * BIO_f_ssl it uses SSL_get_poll_fd; for anything else it will just forward to
+ * BIO_get_fd.
+ */
+__owur int BIO_get_poll_fd(BIO *bio, int *fd); /* BIO_C_GET_POLL_FD */
+__owur int SSL_get_poll_fd(SSL *ssl, int *fd);
+
+/*
+ * Returns the number of milliseconds from now to the earliest (possible) timer
```
Comment (t8m) [inline doc/designs/quic-design/synthesis.md:124]: Is milliseconds the right granularity? And if not, we might need a 64 bit return value type.

File: doc/designs/quic-design/synthesis.md [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+ */
+__owur ssize_t SSL_get_preferred_max_dgram_len(SSL *ssl);
+
+
+/* ========================================================
+ * BIO_dgram enhancements
+ */
+
+/*
+ * Create a BIO_dgram_mem pair. This functions like a BIO pair but
```
Comment (t8m) [inline doc/designs/quic-design/synthesis.md:213]: Is BIO_dgram_mem necessary for MVP?

File: doc/designs/quic-design/synthesis.md [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+ *   Postcondition: n <= num_msg
+ *
+ *   Postcondition: msg[i] for all 0 <= i < n have had their data_len
+ *      modified to the actual amount of data sent or received.
+ *
+ * The flags field of each message defines per-message flags and is used for
+ * both input and output. If a message is processed, this field is written
+ * with any result flags ("message truncated", etc.)
+ *
+ * There are no per-message flags currently defined.
```
Comment (t8m) [inline doc/designs/quic-design/synthesis.md:337]: We might need for example the ECN flag (not for MVP of course).

**Review by t8m — COMMENTED** [review]
Nice work!

File: doc/designs/quic-design/synthesis.md (hunk 2 above)
Comment (hlandau) [reply doc/designs/quic-design/synthesis.md:213]: If we want to support DDD demo 5, it is. It's a common style of existing application.

File: doc/designs/quic-design/synthesis.md (hunk 3 above)
Comment (hlandau) [reply doc/designs/quic-design/synthesis.md:337]: Yeah, definitely.

File: doc/designs/quic-design/synthesis.md [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+    unsigned int    flags;
+} BIO_msg;
+
+#define BIO_WAITFORONE      (1U<<0)
+#define BIO_DONTWAIT        (1U<<1)
+
+__owur ssize_t BIO_sendmmsg(BIO *bio, BIO_msg *msg,
+                            size_t stride, size_t num_msg, unsigned int flags);
+__owur ssize_t BIO_recvmmsg(BIO *bio, BIO_msg *msg,
+                            size_t stride, size_t num_msg, unsigned int flags);
```
Comment (t8m) [inline doc/designs/quic-design/synthesis.md:377]: For future proofing better use uint64_t for both types of flags.

File: doc/designs/quic-design/synthesis.md [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+ * both input and output. If a message is processed, this field is written
+ * with any result flags ("message truncated", etc.)
+ *
+ * There are no per-message flags currently defined.
+ *
+ * The flags argument to these functions defines global flags.
+ *
+ * The defined global flags are:
+ *   BIO_DONTWAIT
+ *      Return immediately if no data is available.
```
Comment (mattcaswell) [inline doc/designs/quic-design/synthesis.md:343]: So default is blocking?

File: doc/designs/quic-design/synthesis.md [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+ *   BIO_WAITFORONE
+ *      Acts as though BIO_DONTWAIT was passed after one datagram has been
+ *      received.
```
Comment (mattcaswell) [inline doc/designs/quic-design/synthesis.md:346]: Why is this needed?

File: doc/designs/quic-design/synthesis.md [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+#define BIO_WAITFORONE      (1U<<0)
+#define BIO_DONTWAIT        (1U<<1)
+
+__owur ssize_t BIO_sendmmsg(BIO *bio, BIO_msg *msg,
+                            size_t stride, size_t num_msg, unsigned int flags);
+__owur ssize_t BIO_recvmmsg(BIO *bio, BIO_msg *msg,
+                            size_t stride, size_t num_msg, unsigned int flags);
+
+/*
+ * (There will of course be BIO_METH_get/set functions corresponding to all of
```
Comment (mattcaswell) [inline doc/designs/quic-design/synthesis.md:380]: typo `BIO_meth_get/set`

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Looks good - but I am concerned by the omission of the SSL_EVENT_CTX stuff. I'm convinced we're going to need it - or something like it.

File: doc/designs/quic-design/synthesis.md (hunk 1 above)
Comment (paulidale) [reply doc/designs/quic-design/synthesis.md:124]: I'd suggest _μs_ to align with `struct timeval`.  On Gbit networks, a _ms_ is a loooong time.  I would entertain argument for/against _ns_ too.

File: doc/designs/quic-design/synthesis.md [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+synthesizes the output of that process with the original QUIC API proposal. The
+the below API has been designed to support the DDD demos with their diffs
+without further changes, and thereby meet the agreed requirements for QUIC.
+
+Response to OTC Questions
+-------------------------
+
+1. *Timer based events*
+
+   This was the subject of extensive discussion in the documents accompanying
```
Comment (paulidale) [inline doc/designs/quic-design/synthesis.md:14]: Could you add a link to these?  I'm not seeing anything relevant in #17991.  Or even add the discussions to #17991 in addition to the code samples.

File: doc/designs/quic-design/synthesis.md (hunk 1 above)
Comment (levitte) [reply doc/designs/quic-design/synthesis.md:124]: A 64-bit time value gives us quite some time regardless...

With *ns*, it wraps after ~584 years (unsigned)
With *µs*, it wraps after ~600000 years (unsigned)

What to choose, what to choose...

Comment (paulidale) [reply doc/designs/quic-design/synthesis.md:124]: _ps_ would give us more room for future performance improvements.
[Plank time](https://en.wikipedia.org/wiki/Planck_units#Planck_time) would be completely future proof but 64 bits wouldn't be adequate.

Comment (t8m) [reply doc/designs/quic-design/synthesis.md:124]: I think ns precision would be purely fictional. So I'd be for μs.

Comment (kroeckx) [reply doc/designs/quic-design/synthesis.md:124]: I think you should go with ns, to match functions like nanosleep() that take a timespec containing ns. It should probably just return a timespec.

File: doc/designs/quic-design/synthesis.md [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+ */
+__owur int BIO_get_next_tick(BIO *bio); /* BIO_C_GET_NEXT_TICK */
+__owur int SSL_get_next_tick(SSL *ssl);
+
+/*
+ * Give libssl an opportunity to do things, for example due to timer events,
+ * even when no specific read or write call is being performed.
+ *
+ * Note that:
+ *   - For TLS this ss a no-op, as there are no timer events for TLS;
```
Comment (kroeckx) [inline doc/designs/quic-design/synthesis.md:138]: ss -> is?

File: doc/designs/quic-design/synthesis.md [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+ * Returns the SSL object's preference for a maximum UDP payload size.
+ *
+ * A client can use this to determine the size of the buffers they should use
+ * for receiving datagrams to be pushed to the SSL object. This is only
+ * necessary if the client is shunting its own buffers to libssl, and only if
+ * support for larger than worst case MTUs is desired. The returned value should
+ * be set as the MTU on the BIO_dgram to indicate to the SSL object that support
+ * for larger MTUs is available.
+ */
+__owur ssize_t SSL_get_preferred_max_dgram_len(SSL *ssl);
```
Comment (kroeckx) [inline doc/designs/quic-design/synthesis.md:205]: I currently fail to see why we need this for an MVP.

File: doc/designs/quic-design/synthesis.md [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+
+Post-MVP Functionality
+----------------------
+
+```c
+/* ========================================================
+ * Alternate API to avoid copying. Not for MVP.
+ */
+int BIO_get_dgram(BIO *bio, void **buf, size_t *buf_len);
+int BIO_return_dgram(BIO *bio, void *buf, size_t buf_len);
```
[the remark this answers is not in the document]
Comment (hlandau) [reply doc/designs/quic-design/synthesis.md:394]: Fair point. This API doesn't need to be finalised until after MVP, though.

File: doc/designs/quic-design/synthesis.md (hunk 8 above)
Comment (kroeckx) [reply doc/designs/quic-design/synthesis.md:14]: I think those demos should end up as part of our documentation, so that this documentation can properly refer to them.

File: doc/designs/quic-design/synthesis.md (hunk 10 above)
Comment (paulidale) [reply doc/designs/quic-design/synthesis.md:205]: Why is this `ssize_t` ?

Comment (hlandau) [reply doc/designs/quic-design/synthesis.md:205]: I guess so that it can fail. But I guess it could return 0 on failure.

File: doc/designs/quic-design/synthesis.md [hunk 12]
```diff
@@ -365,22 +359,46 @@ typedef struct bio_msg_st {
...
 __owur ssize_t BIO_sendmmsg(BIO *bio, BIO_msg *msg,
-                            size_t stride, size_t num_msg, unsigned int flags);
+                            size_t stride, size_t num_msg, uint64_t flags);
 __owur ssize_t BIO_recvmmsg(BIO *bio, BIO_msg *msg,
-                            size_t stride, size_t num_msg, unsigned int flags);
+                            size_t stride, size_t num_msg, uint64_t flags);
 
 /*
- * (There will of course be BIO_METH_get/set functions corresponding to all of
+ * (There will of course be BIO_met_get/set functions corresponding to all of
```
Comment (paulidale) [inline doc/designs/quic-design/synthesis.md:371]: Lost an 'h' in BIO_meth_get/set.

**Review by paulidale — APPROVED** [review]
@openssl/contractors second review?

File: doc/designs/quic-design/synthesis.md (hunk 1 above)
Comment (t8m) [reply doc/designs/quic-design/synthesis.md:124]: This is still unresolved

File: doc/designs/quic-design/synthesis.md [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+ * For QUIC:
+ *   Checks if a new stream has been created by the peer. If it has, creates a
+ *   new SSL object to represent it and returns it. Otherwise, returns NULL.
+ *
+ * For all other methods:
+ *   Returns NULL.
+ *
+ * The flags argument is unused and should be set to zero.
+ *
+ * TODO: Do we want a WAIT flag for synchronous blocking here?
```
Comment (t8m) [inline doc/designs/quic-design/synthesis.md:457]: IMO not necessary for first iteration.

However it would also be nice to have a possibility to set a callback to be called to notify the application that there is a stream to be accepted.

File: doc/designs/quic-design/synthesis.md [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+ */
+SSL *SSL_detach_stream(SSL *ssl);
+
+/*
+ * Attaches a default stream to a QUIC connection object. If the conn object is
+ * not a QUIC connection object, or already has a default stream, this function
+ * fails. The stream must belong to the same connection, or this function fails.
+ *
+ * NOTE: This is not really necessary but I imagine some people might find it
+ *       convenient.
```
Comment (t8m) [inline doc/designs/quic-design/synthesis.md:503]: IMO this is unnecessary and should not be added unless we have a clear use case.

File: doc/designs/quic-design/synthesis.md [hunk 15]
```diff
@@ -0,0 +1,507 @@
...
+__owur int SSL_set_default_stream_type(SSL *ssl, int type);
+
+/*
+ * Detaches a default stream from a QUIC connection object. If the
+ * QUIC connection object does not contain a default stream, returns NULL.
+ * After calling this, calling SSL_get_stream_type on the connection object
+ * returns SSL_STREAM_TYPE_NONE. Always returns NULL for non-QUIC connections.
+ *
+ * NOTE: This is not really necessary but I imagine some people might find it
+ *       convenient.
```
Comment (t8m) [inline doc/designs/quic-design/synthesis.md:493]: Again IMO unnecessary and should not be there without clear use case.

**Comment by paulidale** [comment]
If it's obsoleted, can we close it?
